lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+    return visitOperandMakeRange(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<triton::BroadcastOp>()) {
+    return visitOperandBroadcast(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<triton::SplatOp>()) {
+    return visitOperandSplat(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<triton::ExpandDimsOp>()) {
+    return visitOperandExpandDims(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<triton::AddPtrOp>()) {
+    return visitOperandAddptr(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<arith::ConstantOp>()) {
+    return visitOperandConstSplat(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<arith::RemSIOp>()) {
+    return visitOperandRem(op, state, loc, builder);
+  } else {
+    llvm::dbgs() << "PtrAnalysis: encountered addptr operand produced by an "
+                    "unsupported operation\n";
+    operand.dump();
+    return failure();
+  }
+}
+
+LogicalResult PtrAnalysis::rewriteAddptrOp(triton::AddPtrOp op) {
+  OpBuilder builder(op);
+
+  PtrSState state;
+  if (visitOperandAddptr(op, state, op.getLoc(), builder).failed()) {
+    return failure();
+  }
+
+  knownPtrs[op.getResult()] = state;
+
+  if (op.getPtr().getType().isa<RankedTensorType>()) {
+    auto maketptrOp = state.createTTSMakeTensorPtrOp(builder, op.getLoc());
+    map.map(op.getResult(), maketptrOp.getResult());
+  } else {
+    map.map(op.getResult(), op.getResult());
+  }
+  return success();
+}
+
+LogicalResult PtrAnalysis::rewriteForOp(scf::ForOp op) {
+  SmallVector<Value> newInitArgs;
+
+  SmallVector<std::pair<int, PtrSState>, 5> initArgIndexState;
+  SmallVector<std::pair<int, PtrSState>, 5> knownPtrsTmp;
+
+  llvm::SmallDenseMap<int, PtrSState> initArgIndexMap;
+
+  OpBuilder builder(op);
+
+  // Create a new list of init args
+  for (auto [i, arg] : llvm::enumerate(op.getInitArgs())) {
+    auto mappedV = map.lookupOrNull(arg);
+    PtrSState state;
+
+    if (mappedV) {
+      if (auto makeTPtrOp = mappedV.getDefiningOp<tts::MakeTensorPtrOp>()) {
+        if (visitOperandMakeTPtr(makeTPtrOp, state, op.getLoc(), builder)
+                .failed()) {
+          newInitArgs.push_back(arg);
+          continue;
+        }
+        newInitArgs.push_back(mappedV);
+      } else if (auto addptrOp = mappedV.getDefiningOp<triton::AddPtrOp>()) {
+        assert(!addptrOp.getResult().getType().isa<RankedTensorType>());
+        if (visitOperandAddptr(addptrOp, state, op.getLoc(), builder)
+                .failed()) {
+          newInitArgs.push_back(arg);
+          continue;
+        }
+        newInitArgs.push_back(mappedV);
+      }
+    }
+    // Init arg is not pointer related or prior rewrite has failed. Pass as is
+    else {
+      newInitArgs.push_back(arg);
+      continue;
+    }
+    // Record the PtrSState for later processing
+    initArgIndexState.push_back(std::make_pair(i, state));
+  }
```

Comment:
should we only push if the state is not empty? the only two cases where i think this is necessary are when we call `visitOperandMakeTPtr` or `visitOperandAddptr`

Reply:
We would only hit this statement if it is a pointer that we have successfully analyzed. If it is not pointer related, or if the analyze failed, we would not have pushed into `initArgIndexState`. Does that answer your question?

Reply:
This makes sense now. Thank you.